drivers/sensor/sgp40/sgp40.c

```diff
@@ -0,0 +1,309 @@
+		data->t_param[0], data->t_param[1], data->t_param[2],
+	};
+
+	return i2c_write(cfg->bus, tx_buf, sizeof(tx_buf),
+			 cfg->i2c_addr);
+}
+
+/*
+ * Set the params for RH/T compensation
+ */
+static int sgp40_update_compensation_params(const struct device *dev,
+			   int16_t t, int16_t rh)
+{
+	struct sgp40_data *data = dev->data;
+	int16_t rh_ticks, t_ticks;
+
+	sgp40_clamp(rh, SGP40_COMP_MIN_RH, SGP40_COMP_MAX_RH);
+	sgp40_clamp(t, SGP40_COMP_MIN_T, SGP40_COMP_MAX_T);
+
+	/*
+	 * Temperature and RH conversion to ticks as explained in datasheet
+	 * in section "I2C commands"
+	 */
+	rh_ticks = (rh * 0xFFFF / 100U) + 0.5;
+	t_ticks = ((t + 45U) * 0xFFFF / 175U) + 0.5;
+
+	data->rh_param[0] = rh_ticks >> 8;
+	data->rh_param[1] = rh_ticks & 0xFF;
+	data->rh_param[2] = sgp40_compute_crc(rh_ticks);
+
+	data->t_param[0] = t_ticks >> 8;
+	data->t_param[1] = t_ticks & 0xFF;
+	data->t_param[2] = sgp40_compute_crc(t_ticks);
+
+	return 0;
+}
+
+/*
+ * Public API for setting compensation data
+ */
+int sgp40_update_compensation_data(const struct device *dev,
+		struct sgp40_compensation_data comp_data)
+{
+	return sgp40_update_compensation_params(dev, comp_data.t, comp_data.rh);
+}
+
+/*
+ * Public API for "disabling" T/RH compensation
+ */
+int sgp40_disable_compensation(const struct device *dev)
+{
+	return sgp40_update_compensation_params(dev, 25, 50);
+}
+
+static int sgp40_selftest(const struct device *dev)
+{
+	const struct sgp40_config *cfg = dev->config;
+	uint8_t rx_buf[3];
+	uint16_t raw_sample;
+
+	if (sgp40_write_command(dev, SGP40_CMD_MEASURE_TEST)) {
+		LOG_ERR("Failed to start selftest!");
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(SGP40_TEST_WAIT_MS));
+
+	if (i2c_read(cfg->bus, rx_buf, sizeof(rx_buf), cfg->i2c_addr)) {
+		LOG_ERR("Failed to read data sample.");
+		return -EIO;
+	}
+
+	raw_sample = sys_get_be16(rx_buf);
+	if (sgp40_compute_crc(raw_sample) != rx_buf[2]) {
+		LOG_ERR("Received invalid CRC from selftest.");
+		return -EIO;
+	}
+
+	if (raw_sample != SGP40_TEST_OK) {
+		LOG_ERR("Selftest failed.");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int sgp40_attr_set(const struct device *dev, enum sensor_channel chan,
+			   enum sensor_attribute attr,
+			   const struct sensor_value *val)
+{
+	return -ENOTSUP;
+}
+
+static int sgp40_sample_fetch(const struct device *dev,
+			       enum sensor_channel chan)
+{
+	struct sgp40_data *data = dev->data;
+	const struct sgp40_config *cfg = dev->config;
+
+	uint8_t rx_buf[3];
+	uint16_t raw_sample;
+
+	if (chan != SENSOR_CHAN_GAS_RES && chan != SENSOR_CHAN_ALL) {
+		return -ENOTSUP;
+	}
+
+	if (sgp40_start_measurement(dev)) {
+		LOG_ERR("Failed to start measurement.");
+		return -EIO;
+	}
+
+	k_sleep(K_MSEC(SGP40_MEASURE_WAIT_MS));
+
+	if (i2c_read(cfg->bus, rx_buf, sizeof(rx_buf), cfg->i2c_addr)) {
+		LOG_ERR("Failed to read data sample.");
+		return -EIO;
+	}
+
+	raw_sample = sys_get_be16(rx_buf);
+ 	if (sgp40_compute_crc(raw_sample) != rx_buf[2]) {
```

Comment:
ditto, error code propagation

Reply:
probably because of cargo-cult programming. i am new to zephyr..
i have removed them.

Reply:
ditto